[RHCLOUD-15003] Feature migrations

MikelAlejoBR opened the pull request:
- Uses `golang-migrate`(https://github.com/golang-migrate)
- Creates a new `schema_migrations_go` table to store the 
- Migrates all the `sources-api/db/migrate/*.rb` files to SQL files
    -  Replaces the `yyyy-MM-ddHHMMSS_${description}` format with a sequential `1_${description}` one. The reason for this is that `golang-migrate` stores the versions as an `int8` instead of the `varchar`s that Rails stores. So to make the table contents easier to read, instead of storing gigantic integers, simple-small numbers are stored.
    - Also, in order not to incur in "already exists" errors, some safeguards have been implemented. These check on the prior existence of the tables and foreign key constraints before creating the tables. This way, if the first migration is applied in an already populated database, no error will be thrown, and nothing will be done.

## Links

[RHCLOUD-15003](https://issues.redhat.com/browse/RHCLOUD-15003)

MikelAlejoBR committed:
feature: set up "gomigrate" to perform migrations programatically

MikelAlejoBR committed:
feature: add initial migration from the current state of the db

This migration is a snapshot of the current schema of the database. In
order not to make it clash with it when it gets applied, some safeguards
are implemented which check the existence of the tables and the
corresponding foreign keys.

Therefore, the tables only get created when they don't exist, and the
same thing goes with the foreign keys.

Althought not everything gets checked: the sequences and constraints
associated with a table are assumed to be non existent if the table
itself doesn't exist.

MikelAlejoBR committed:
feature: use gormigrate for managing migrations

lindgrenj6 commented on dao/db.go:
```diff
@@ -41,6 +44,19 @@ func Init() {
+		DB.Exec(`CREATE SCHEMA "public"`)
+	}
+
+	if conf.MigrationsSetup || conf.MigrationsReset {
```
I think I want the `setup` flag to just create the db - kinda like we do for `sources_api_test`. 

I want the migrations to just always run every time - since the app theoretically shouldn't be running if the db isnt migrated completely. This also prevents us from having to have an initContainer to run the migrations.

MikelAlejoBR committed:
refactor: the setup command should only create the database

Mimicking the behaviour of the similar "createDb" command in the tests,
the "setup" command should only try to create the DB and exit.

MikelAlejoBR committed:
refactor: improve error handling when resetting the database

MikelAlejoBR committed:
refactor: exit after successfully resetting the schema

MikelAlejoBR committed:
feature: implement locking mechanism for migrations

MikelAlejoBR committed:
refactor: use "SETEX" instead of "SETNX"

Our Redis instance has the version 5.0, so we can't use "SETNX" yet.

lindgrenj6 commented on main.go:
```diff
@@ -19,9 +21,19 @@ var conf = config.Get()
+	setUpDatabase := flag.Bool("setup", false, "create the schema and run the migrations")
+	resetDatabase := flag.Bool("reset", false, "drop the schema, recreate it and run the migrations")
+
+	flag.Parse()
```
Final comment: we can move these over to the `config/` package here: https://github.com/RedHatInsights/sources-api-go/blob/62ea4f6533db5dbddb5d5dba4682f6045eca432b/config/config.go#L111

That way they only get parsed once and the config struct has them (like you do below).

MikelAlejoBR replied:
Done!

lindgrenj6 reviewed (approved):
Ok this looks good now. I just have one tiny comment moving the flags to the config package since I was trying to get all of our runtime configuration over there. 

---

@lpichler I am having Mikel split this into multiple PRs so we can get it merged so we can (hopefully) do our first migration over here on the Go side for the `org_id` migration. With the PR in its current state we'll temporarily have 2 migration tables, one for Go and one for Rails. 

After we get done removing the Rails app completely we'll add in the migrations from #175 to remove the rails migration AND rename all the foreign keys from the rails generated to more human readable names. 

---

Please test locally - I did and it appears that it didn't effect any of the current tables than creating a `migrations` table that stores the migrations that have been ran.

MikelAlejoBR committed:
refactor: parse the flags on the config file

lpichler commented on dao/db.go:
```diff
@@ -46,8 +46,15 @@ func Init() {
 	if conf.MigrationsReset {
-		DB.Exec(`DROP SCHEMA "public" CASCADE`)
-		DB.Exec(`CREATE SCHEMA "public"`)
+		err = DB.Exec(`DROP SCHEMA "public" CASCADE`).Error
```
you have in setup drop database  - Should not be here also dropping database ?

lpichler commented on dao/db.go:
```diff
@@ -46,8 +46,15 @@ func Init() {
+			logging.Log.Fatalf(`Error dropping the "public" schema: %s`, err)
+		}
+
+		err = DB.Exec(`CREATE SCHEMA "public"`).Error
```
same here

MikelAlejoBR committed:
feature: drop the database entirely when receiving "-reset" flag

lpichler commented on dao/db.go:
```diff
@@ -41,6 +45,61 @@ func Init() {
+		}
+
+		// Log and exit so that the application can be rerun without the "reset" flag.
+		logging.Log.Info(`The schema "public" has been reset`)
```
yes it is true but regard to last changes - I am suggesting change it to "...database XX has been reset"

lpichler commented on config/config.go:
```diff
@@ -109,13 +111,17 @@ func Get() *SourcesApiConfig {
 	// Parse any Flags (using our own flag set to not conflict with the global flag)
 	fs := flag.NewFlagSet("runtime", flag.ContinueOnError)
 	availabilityListener := fs.Bool("listener", false, "run availability status listener")
+	setUpDatabase := fs.Bool("setup", false, "create the schema and run the migrations")
```
schema -> database

lpichler commented on config/config.go:
```diff
@@ -109,13 +111,17 @@ func Get() *SourcesApiConfig {
 	fs := flag.NewFlagSet("runtime", flag.ContinueOnError)
 	availabilityListener := fs.Bool("listener", false, "run availability status listener")
+	setUpDatabase := fs.Bool("setup", false, "create the schema and run the migrations")
+	resetDatabase := fs.Bool("reset", false, "drop the schema, recreate it and run the migrations")
```
schema -> database

MikelAlejoBR committed:
refactor: change the log messages to a more accurate ones

MikelAlejoBR committed:
fix: not able to connect to reset or setup the database

When the database didn't exist, the service was still trying to connect
to a non-existing database, which would make the program exit and not
even reach the "reset" or "setup" code.

In order to avoid that problem, if either of those flags is passed, a
connection is opened to the maintenance "postgres" database, to be able
to perform those operations, since it is not possbile to drop a database
from whithin the connection that is connected to it.

MikelAlejoBR committed:
feature: include info messages on migration start and end
